base_substate/models/base_substate_mixin.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2019 Akretion
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models, _
+from odoo.exceptions import ValidationError
+
+
+class BaseSubstateMixin(models.AbstractModel):
+    _name = "base.substate.mixin"
+    _description = 'BaseSubstate Mixin'
+    # _inherit = ['mail.thread']
+
+    def _get_default_substate_id(self, state_val=False):
+        """ Gives default substate_id """
+        search_domain = self._get_default_substate_domain(state_val)
+        # perform search, return the first found
+        return self.env['base.substate'].search(search_domain,
+                                                order='sequence', limit=1).id
+
+    def _get_default_substate_domain(self, state_val=False):
+        """ Override this method
+            to change domain values
+        """
+        if not state_val:
+            state_val = self._get_default_state_value()
+        substate_type = self._get_substate_type()
+        state_field = substate_type.target_state_field
+        if self and not state_val and state_field in self._fields:
+            state_val = self[state_field]
+
+        domain = [('target_state_value_id.target_state_value', '=', state_val)]
+        domain += [('target_state_value_id.base_substate_type_id',
+                    '=', substate_type.id)]
+        return domain
+
+    def _get_default_state_value(self,):
+        """ Override this method
+            to change state_value
+            """
+        return 'draft'
+
+    def _get_substate_type(self,):
+        """ Override this method
+            to change substate_type (get by xml id for example)
+        """
+        return self.env['base.substate.type'].search(
+            [('model', '=', self._name)], limit=1)
+
+    substate_id = fields.Many2one(
+        'base.substate',
+        string='Sub State',
+        ondelete='restrict',
+        default=_get_default_substate_id,
```

Comment:
```suggestion
        default=lambda self: self._get_default_substate_id(),
```
then move this field above the methods